crypto/data_formats.go

```diff
@@ -15,41 +15,72 @@
+
+	pb "github.com/google/trillian"
 )
 
 // This file contains struct specific mappings and data structures.
-// TODO(gdbelvin): remove data-structure specific operations.
 
-// Constants used as map keys when building input for ObjectHash. They must not be changed
-// as this will change the output of hashRoot()
+type logVersion int
+
+const (
+	// LogRootV0 indicates version 0 for the canonical representation of the SignedLogRoot.
+	LogRootV0 logVersion = iota
+)
+
+// CanonicalLogRoot returns a canonical TLS serialization of the log root.
+func CanonicalLogRoot(r *pb.SignedLogRoot, version logVersion) ([]byte, error) {
+	switch version {
+	case LogRootV0:
+		root := struct {
+			RootHash       []byte `tls:"minlen:0,maxlen:128"`
+			TimestampNanos uint64
+			TreeSize       uint64
+			LogID          uint64
+		}{
+			RootHash:       r.RootHash,
+			TimestampNanos: uint64(r.TimestampNanos),
+			TreeSize:       uint64(r.TreeSize),
+			LogID:          uint64(r.LogId),
+		}
+		return tls.Marshal(root)
+	default:
+		return nil, fmt.Errorf("CanonicalLogRoot(): Unknown version: %v", version)
+	}
+}
+
+type mapVersion int
+
 const (
-	mapKeyRootHash       string = "RootHash"
-	mapKeyTimestampNanos string = "TimestampNanos"
-	mapKeyTreeSize       string = "TreeSize"
+	// MapRootV0 indicates version 0 for the canonical representation of the SignedMapRoot.
+	MapRootV0 mapVersion = iota
 )
 
-// HashLogRoot hashes SignedLogRoot objects using ObjectHash with
-// "RootHash", "TimestampNanos", and "TreeSize", used as keys in
-// a map.
-func HashLogRoot(root trillian.SignedLogRoot) ([]byte, error) {
-	// Pull out the fields we want to hash.
-	// Caution: use string format for int64 values as they can overflow when
-	// JSON encoded otherwise (it uses floats). We want to be sure that people
-	// using JSON to verify hashes can build the exact same input to ObjectHash.
-	rootMap := map[string]interface{}{
-		mapKeyRootHash:       base64.StdEncoding.EncodeToString(root.RootHash),
-		mapKeyTimestampNanos: strconv.FormatInt(root.TimestampNanos, 10),
-		mapKeyTreeSize:       strconv.FormatInt(root.TreeSize, 10)}
-
-	hash, err := objecthash.ObjectHash(rootMap)
-	if err != nil {
-		return nil, fmt.Errorf("ObjectHash(%#v): %v", rootMap, err)
+// CanonicalMapRoot returns a canonical TLS serialization of the map root.
+func CanonicalMapRoot(r *pb.SignedMapRoot, version mapVersion) ([]byte, error) {
+	switch version {
+	case MapRootV0:
+		root := struct {
+			RootHash       []byte `tls:"minlen:0,maxlen:128"`
+			TimestampNanos uint64
+			MapID          uint64
+			MapRevision    uint64
+			MetadataType   []byte `tls:"minlen:0,maxlen:65535"`
+			MetadataValue  []byte `tls:"minlen:0,maxlen:65535"`
+		}{
+			RootHash:       r.RootHash,
+			TimestampNanos: uint64(r.TimestampNanos),
+			MapID:          uint64(r.MapId),
+			MapRevision:    uint64(r.MapRevision),
+			MetadataType:   []byte(r.Metadata.GetTypeUrl()),
+			MetadataValue:  r.Metadata.GetValue(),
```

Comment:
I don't know (yet) anything about Map metadata so I dunno if this is sensible or not...
Summoning @AlCutter ?

Reply:
I don't think we know yet, I was vaguely assuming that it should probably be an opaque `[]byte{}` for both Log and Map and the contents would be defined by the personality layer, having the `MetadataType` might not be strictly necessary if that is indeed how it ends up rolling, but doesn't necessarily seem like a bad either assuming that the Trillian roots only go as far as the personalities (otherwise we'll be adding a fair number of bytes overhead to each root...)